### ISSUE TITLE
Build fails on MacOS with gcc-10.3.0

### ISSUE THREAD

**Issue by gambiteer** [gambit/gambit#681]
I don't know why this is suddenly happening:
```
>>>> building module _define-library for C
/bin/sh: line 1: 59292 Illegal instruction: 4  DYLD_LIBRARY_PATH=../../lib:../../gsi:../../gsc: ../../gsc/gsc -:~~bin=../../bin,~~lib=../../lib,~~include=../../include -f -e "(let ((x (##build-module \"$main_mod\" (quote $target) (quote ((module-ref $modref)))))) (if (not x) (begin (println \"*** module failed to build\") (exit 1))))"
>>>> building module _define-library/define-library-expand for C
/bin/sh: line 1: 59293 Illegal instruction: 4  DYLD_LIBRARY_PATH=../../lib:../../gsi:../../gsc: ../../gsc/gsc -:~~bin=../../bin,~~lib=../../lib,~~include=../../include -f -e "(let ((x (##build-module \"$main_mod\" (quote $target) (quote ((module-ref $modref)))))) (if (not x) (begin (println \"*** module failed to build\") (exit 1))))"
>>>> building module _define-library/debug for C
/bin/sh: line 1: 59294 Illegal instruction: 4  DYLD_LIBRARY_PATH=../../lib:../../gsi:../../gsc: ../../gsc/gsc -:~~bin=../../bin,~~lib=../../lib,~~include=../../include -f -e "(let ((x (##build-module \"$main_mod\" (quote $target) (quote ((module-ref $modref)))))) (if (not x) (begin (println \"*** module failed to build\") (exit 1))))"
make[2]: *** [modules-pre] Error 132
make[1]: *** [_define-library] Error 1
make: *** [modules] Error 2
```
The configure is
```
[Bradleys-Mac-mini:~/programs/gambit] lucier% gsi -v
v4.9.3-1370-g66faadf8 20210409193845 x86_64-apple-darwin19.6.0 "./configure 'CC=gcc-10 -march=native -D___CAN_IMPORT_CLIB_DYNAMICALLY' '--enable-single-host' '--enable-shared' '--enable-multiple-versions'"
```
GCC is
```
[Bradleys-Mac-mini:~/programs/gambit] lucier% gcc-10 -v
Using built-in specs.
COLLECT_GCC=gcc-10
COLLECT_LTO_WRAPPER=/opt/sw/lib/gcc10/libexec/gcc/x86_64-apple-darwin19.6.0/10.3.0/lto-wrapper
Target: x86_64-apple-darwin19.6.0
Configured with: ../gcc-10.3.0/configure --prefix=/opt/sw --prefix=/opt/sw/lib/gcc10 --mandir=/opt/sw/share/man --infodir=/opt/sw/lib/gcc10/info --enable-languages=c,c++,fortran,lto,objc,obj-c++ --with-gmp=/opt/sw --with-libiconv-prefix=/opt/sw --with-isl=/opt/sw --with-mpc=/opt/sw --with-system-zlib --program-suffix=-fsf-10 --disable-multilib --enable-stage1-checking --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk --with-pkgversion='Fink gcc10 10.3.0-1' --with-bugurl=https://github.com/fink/fink-distributions/issues
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 10.3.0 (Fink gcc10 10.3.0-1) 
```
UPDATE: It's a `gcc` problem, it builds fine with
```
[Bradleys-Mac-mini:~/programs/gambit] lucier% gsi -v
v4.9.3-1371-g1dbeaabd 20210413203916 x86_64-apple-darwin19.6.0 "./configure 'CC=gcc-fsf-9 -march=native -D___CAN_IMPORT_CLIB_DYNAMICALLY' '--enable-single-host' '--enable-shared' '--enable-multiple-versions'"
```
and it built fine with GCC 10.2.0.  Blah

Any suggestions about how I might debug this to report it to the GCC folks?

**Comment by gambiteer** [comment]
This does not seem to be a Gambit bug.

**Comment by feeley** [comment]
Did you find the problem?

**Comment by gambiteer** [comment]
No, there are too many things to be checked easily, and I don't really understand the error message.  Do you?

Perhaps I'll start by removing the "-march=native".

**Comment by gambiteer** [comment]
Still happens without `-march=native`.
I don't understand the module building system.  Where is the message coming from?

**Comment by gambiteer** [comment]
And, same problem with `CC=gcc-10`, without ` -D___CAN_IMPORT_CLIB_DYNAMICALLY`.

**Comment by gambiteer** [comment]
And, it fails with a generic build of gcc-10.3.0, not just with the Fink build.

**Comment by gambiteer** [comment]
Adding  `-mmacosx-version-min=10.15` (I'm running on Catalina) doesn't fix it, see https://stackoverflow.com/questions/14268887/what-is-the-illegal-instruction-4-error-and-why-does-mmacosx-version-min-10 .

**Comment by gambiteer** [comment]
It's crashing at line 3280 of `c_intf.c`.  More to come.

**Comment by gambiteer** [comment]
I don't know how to run the failing command
```
>>>> building module _define-library for C
/bin/sh: line 1: 98079 Illegal instruction: 4  (core dumped) DYLD_LIBRARY_PATH=../../lib:../../gsi:../../gsc: ../../gsc/gsc -:~~bin=../../bin,~~lib=../../lib,~~include=../../include -f -e "(let ((x (##build-module \"$main_mod\" (quote $target) (quote ((module-ref $modref)))))) (if (not x) (begin (println \"*** module failed to build\") (exit 1))))"
```
by hand in the debugger, because I don't know what `$main_mod`, `$target`, and `$modref` are supposed to be to build `_define-library`.

**Comment by gambiteer** [comment]
The offending instruction is
```
->  0x103d60010 <+1520>: movl   (%rbp,%r10,4), %esi
```
I have no frigging clue what about this instruction it doesn't like.

**Comment by feeley** [comment]
Indeed that is strange.  For the `$main_mod` etc you can add a trace in `lib/module-common.mk` specifically in this part:

**Comment by gambiteer** [comment]
I set up a GCC PR:

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100152

I think I'll wait a bit to see what they come up with.

**Comment by gambiteer** [comment]
It has nothing to do with building modules, it's very simple to reproduce (at least on my machine), see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100152#c16

**Comment by gambiteer** [comment]
Marc:  I configured with
```
 ./configure 'CC=/usr/local/gcc-10.3.0/bin/gcc -save-temps -g -W -Wall' --enable-single-host --enable-shared
```
i.e., I turned on a lot of warnings, including ones that aren't usually on to see perhaps whether `gcc-10.3.0` has treated some fishy C construct in the Gambit sources differently from how `gcc-10.2.0` treats it.  Here's the build log, it would help if you could see whether any of the items it flags are a real problem.
I'm a bit concerned because Iain Sandoe has been putting a lot of effort into reproducing the bug (including building Gambit on his own machine) and I'd like to make sure it isn't a newly illuminated problem in Gambit's source code.

[build.log](https://github.com/gambit/gambit/files/6353567/build.log)

**Comment by feeley** [comment]
The log sees fine (innocuous warnings at least for c_intf.c).  The fallthroughs are intentional.

**Comment by gambiteer** [comment]
It appears that Iain has analyzed the problem, but not yet a fix.  (But he works really fast. ;-)
The code generation bug is tickled on Darwin with shared libraries with optimization level -O2 and above.
So it would not appear if c_intf.c et al. were compiled with -O1 instead of -O3.

**Comment by feeley** [comment]
A fix that avoids the gcc 10.3 bug is 920c16b81151a88151a17f274b816c5297f92ee8.
